contracts/src/v0.8/workflow/dev/WorkflowRegistryManager.sol

```diff
@@ -0,0 +1,221 @@
+    return activeVersionNumber;
+  }
+
+  /// @notice Retrieves the version number of the latest registered WorkflowRegistry version.
+  /// @return latestVersionNumber The version number of the latest version.
+  /// @custom:throws NoVersionsRegistered if s_latestVersionNumber is 0.
+  function getLatestVersionNumber() external view returns (uint32 latestVersionNumber) {
+    latestVersionNumber = s_latestVersionNumber;
+    if (latestVersionNumber == 0) revert NoVersionsRegistered();
+    return latestVersionNumber;
+  }
+
+  // ================================================================
+  // |                          PRIVATE                             |
+  // ================================================================
+  /// @dev This function deactivates the currently active version (if any) before activating the specified version. It
+  /// emits events for both deactivation and activation.
+  /// @param versionNumber The version number of the version to activate.
+  /// @custom:throws IndexOutOfBounds if the version number does not exist.
+  function _activateVersion(uint32 versionNumber) private {
+    // Cache the current active version number to reduce storage reads
+    uint32 currentActiveVersionNumber = s_activeVersionNumber;
+
+    // Check that the provided version number is within a valid range
+    if (versionNumber == 0 || versionNumber > s_latestVersionNumber) {
+      revert VersionNotRegistered(versionNumber);
+    }
+
+    // Emit deactivation event if there is an active version
+    if (currentActiveVersionNumber != type(uint32).max) {
+      Version memory currentActive = s_versions[currentActiveVersionNumber];
+      emit VersionDeactivatedV1(currentActive.contractAddress, currentActive.chainID, currentActiveVersionNumber);
+    }
+
+    // Set the new active version (which deactivates the previous one)
+    s_activeVersionNumber = versionNumber;
+    Version memory newActive = s_versions[versionNumber];
+    emit VersionActivatedV1(newActive.contractAddress, newActive.chainID, versionNumber);
+  }
+
+  /// @dev Validates that a given contract address is non-zero, contains code, and supports the IWorkflowRegistry interface.
+  /// @param contractAddress The address of the contract to validate.
+  /// @custom:throws InvalidContractAddress if the address is zero or contains no code.
+  /// @custom:throws InvalidContractType if the contract does not implement typeAndVersion().
+  function _getTypeAndVersionForContract(address contractAddress) internal view returns (string memory) {
+    if (!_isNonZeroWithCode(contractAddress)) {
+      revert InvalidContractAddress(contractAddress);
+    }
+
+    try ITypeAndVersion(contractAddress).typeAndVersion() returns (string memory retrievedVersion) {
+      return retrievedVersion;
+    } catch {
+      revert InvalidContractType(contractAddress);
+    }
+  }
+
+  function _isNonZeroWithCode(address _addr) internal view returns (bool) {
```

Comment:
so for small functions like these, does using `--optimize` automatically inline them or would you say it's still better to write it manually? I thought this would be better for readability purposes and more closer to writing normal non solidity code, and it's kind of vague on how much optimization a newer version provides I guess. I was planning on trying out various versions and running gas reports on them to see but theoretically?

Reply:
> does using --optimize automatically inline them 

You should assume the compiler is horrible and does nothing for you, because often that's the case. 

I agree that for normal code it could totally make sense to have single line functions just for readability